Tutorial for sampling events of an energy-dependent temporal model

fabiopintore opened the pull request:
This PR aims at create a tutorial that will show how to sample events from a source with an energy-dependent temporal model. The tutorial describes how to generate such a temporal model, how to sample events from it and then it presents a fit of a given time interval. Results are compared with the simulated ones.

AtreyeeS commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+
+bkg_model = FoVBackgroundModel(dataset_name="my-dataset")
+
+model_fit = [model, bkg_model]
```
Is it a good idea to fit the background within 100 sec bins? @registerrier ?

fabiopintore replied:
This is certainly not a good estimate of the bkg, in general. However, in this case we are assuming to know it very well (it is read from the IRF) and the bkg normalization is compatible with 1. So, it is not the best approach but does not introduce a large bias.

AtreyeeS replied:
I dont know what would be the best to show in a tutorial. The background model should be shared between all the split bins, so they are not really independent... In practice, what might be good is to first fit the background on the observation, masking the source. Then split the observation into smaller bins, and keep the parameters of the background model fixed to the previously obtained value. This would be too complex (and unnecessary) to show in this tutorial.

I am just a bit afraid that people trying to fit the background model on very small time scales may end up with nonsense results

AtreyeeS replied:
Maybe, rather than get into all this, we can use the `LightCurveEstimator` for 3 energy bins which can show the spectral evolution.

fabiopintore replied:
yes, probably the approach you propose to estimate the BKG is too much for the aims of the tutorial. What about if we just write a warning to the reader that this is only a very simplified example and that the bkg estimate might be misleading using the method we show?

AtreyeeS commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+# ---------
+#
+# -  Try to create a temporal model with a more complex energy-dependent
+#    evolution;
```
Maybe suggest to read the template model in gammapy data and perform a simulation?

fabiopintore replied:
Ok.

AtreyeeS reviewed (approved):
Thank you @fabiopintore for the excellent tutorial.
I left some small inline comments 

I guess we need to specify the thumbnail and tooltip. Is the instruction in the dev docs,  https://docs.gammapy.org/dev/development/doc_howto.html?highlight=tooltip#choose-a-thumbnail-and-tooltip-for-the-tutorial-gallery , still valid in the new scheme of .py files @adonath ?

adonath commented:
@AtreyeeS No, this is not valid anymore. Instead one can use https://sphinx-gallery.github.io/stable/auto_examples/plot_4_choose_thumbnail.html 

We should adapt the dev docs...

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+"""
+Sample a source with energy-dependent evolution
```
Ok.

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+-------------
+
+To understand how to generate a model and a MapDataset and how to fit the data, please refer to the `~gammapy.modeling.models.SkyModel` and
+:doc:`/tutorials/analysis-3d/simulate_3d` tutorial. To know how to sampling events for more standards sources, we suggest to visit the event sampler :doc:`/tutorials/analysis-3d/event_sampling` tutorial.
```
Ok.

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+time_min = np.arange(0, 3600, time_bins) * u.s
+time_max = np.arange(720, 4320, time_bins) * u.s
+edges = np.append(time_min, time_max[-1])
+time_axis = MapAxis.from_edges(edges=edges, name="time", interp="lin")
```
What do you mean? That the number of time bins is too small?

AtreyeeS replied:
No no.. just that you do not need to explicitly create the time min and max values. Gammapy can create the bins automatically if you pass the min, max and the number of bins. The 4 lines of code here can be reduced to one

fabiopintore replied:
Ah, right. I've simplified it using `from_bounds`.

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+#
+
+# make an array with the time and energy shape
+data = np.ones((len(time_min), energy_axis.nbin)) * u.cm**-2 * u.s**-1 * u.TeV**-1
```
Ok.

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+# Now, we define `LightCurveTemplateTemporalModel` passing it the
+# map we created before. We write it on disk and then we read it
+# again with `LightCurveTemplateTemporalModel.read`. When the model
+# is from a map, the arguments `format="map"` is mandatory.
```
Done.

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+
+######################################################################
+# We note that an interpolation scheme is also provided when loading
+# an map: for a an energy-dependent temporal model, the `method` and
```
Ok.

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+# an map: for a an energy-dependent temporal model, the `method` and
+# `values_scale` arguments by default are set to `linear` and `log`.
+# We warn the reader to carefully check the interpolation method used
+# for the time axis, as different methods provide different results.
```
Ok.

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+# `values_scale` arguments by default are set to `linear` and `log`.
+# We warn the reader to carefully check the interpolation method used
+# for the time axis, as different methods provide different results.
+# In general, we suggest `linear` interpolation for the time, `log`
```
Ok.

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+temporal_model.values_scale = "log"  # default
+
+######################################################################
+# We can have a visual inspection of the temporal model at 1 TeV:
```
Ok.

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+
+t = temporal_model.reference_time + np.linspace(-100, 3600, 10000) * u.s
+t = Time(t, format="mjd")
+f = temporal_model.evaluate(t, energy=1 * u.TeV)
```
Right, I've removed it.

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+# We can have a visual inspection of the temporal model at 1 TeV:
+#
+
+t = temporal_model.reference_time + np.linspace(-100, 3600, 10000) * u.s
```
Ok.

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+#
+
+t = temporal_model.reference_time + np.linspace(-100, 3600, 10000) * u.s
+t = Time(t, format="mjd")
```
Removed

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+energy_axis_true = MapAxis.from_energy_bounds(
+    "0.05 TeV", "150 TeV", nbin=10, per_decade=True, name="energy_true"
+)
+migra_axis = MapAxis.from_bounds(0.5, 2, nbin=150, node_type="edges", name="migra")
```
In the event sampling tutorial, a `migra_axis` is created. It's not so crucial, but I think it can be useful in this tutorial as well.

AtreyeeS replied:
Do we need a finer migration bin for the event sampling than we have in the default? Why?

fabiopintore replied:
Well, for event sampler, generally speaking the finer the better...

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+
+src_events = events.select_region(on_region)
+
+src_events.peek()
```
Ok.

fabiopintore commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,384 @@
+
+######################################################################
+# The best-fit spectral parameters are similar but not exactly equal
+# to the simulated ones. This is due to interpolation method applied
```
Ok, I've adopted your sentence

fabiopintore commented:
Thanks @AtreyeeS , I've implemented all your suggestions. I've also added a description of the energy-dependent temporal model map. Let me know if this is OK for you.

QRemy commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,413 @@
+
+
+######################################################################
+# Create the energy-dependent model
```
```suggestion
# Create the energy-dependent temporal model
```

QRemy commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,413 @@
+# The source to simulate has a spectrum that varies as a function of
+# the time.
+# We define its spatial morphology as `point-like`: we note that this
+# is a mandatory condition to simulate energy-dependent sources! Other,
```
```suggestion
# is a mandatory condition to simulate energy-dependent temporal model. Other,
```

QRemy commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,413 @@
+# the time.
+# We define its spatial morphology as `point-like`: we note that this
+# is a mandatory condition to simulate energy-dependent sources! Other,
+# extended morphologies will raise an error!
```
Is there a way or motivation to support that in future ? I guess no, then maybe we can justify this design choice here ?

QRemy commented on examples/tutorials/analysis-3d/event_sampling_nrg_depend_models.py:
```diff
@@ -0,0 +1,413 @@
+
+temporal_model = LightCurveTemplateTemporalModel.read(filename, format="map")
+
+######################################################################
+# We note that an interpolation scheme is also provided when loading
+# a map: for a an energy-dependent temporal model, the `method` and
+# `values_scale` arguments by default are set to `linear` and `log`.
+# We warn the reader to carefully check the interpolation method used
+# for the time axis while creating the template model, as different
+# methods provide different results.
+# By default, we assume `linear` interpolation for the time, `log`
+# for the energies and values.
+# Users can modify the `method` and `values_scale` arguments but we
+# warn that this should be done only when the users knows the consequences
+# of the changes. Here, we show how to set them explicitely:
+#
+
+temporal_model.method = "linear"  # default
+temporal_model.values_scale = "log"  # default
```
maybe do not show that

QRemy reviewed (changes requested):
Thanks @fabiopintore, I have some doubts about the analysis part. I think you shoud save the simulated events in an observation and then read it for the analysis. You could show a 1d analysis so you don't have the problem of the background fitting. It could include a loop over time bins that apply Observation.select_time(), run the data reduction and perform the fit. Finally show the fitted norm, index versus time.

fabiopintore commented:
Hi @registerrier , unfortunately I haven't done much in the last 10 days. 
So the tutorial is not updated and I haven't yet included @QRemy comments...

registerrier commented:
> Hi @registerrier , unfortunately I haven't done much in the last 10 days. So the tutorial is not updated and I haven't yet included @QRemy comments...

HI @fabiopintore , I will provide an updated and rebased version at the end of the afternoon. 
If it is OK with you I will push directly to your branch.

registerrier commented:
The modification suggested have been taken into account and a new PR based on this branch has been opened that supersedes this PR. Closing now. 
Thanks @fabiopintore !
